### ISSUE TITLE
[no-unused-var] False positive for nested namespace declaration

### ISSUE THREAD

**Issue by hiranya911** [typescript-eslint/typescript-eslint#2573]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```js
{
  root: true,
  parser: '@typescript-eslint/parser',
  plugins: [
    '@typescript-eslint',
  ],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
  ],
  rules: {
    // Following checks are temporarily disabled. We shall incrementally enable them in the
    // future, fixing any violations as we go.
    '@typescript-eslint/no-non-null-assertion': 0,

    // Disabled checks
    '@typescript-eslint/no-explicit-any': 0,
    '@typescript-eslint/no-use-before-define': 0,

    // Required checks
    'indent': ['error', 2],
    'keyword-spacing': ['error'],
    'max-len': [
      'error',
      {
        'code': 120,
        'ignoreUrls': true
      }
    ],
    "object-curly-spacing": [2, "always"],
    '@typescript-eslint/explicit-function-return-type': [
      'error',
      {
        'allowExpressions': true,
        'allowTypedFunctionExpressions': true,
        'allowHigherOrderFunctions': true
      }
    ],
    'no-unused-vars': 'off', // Must be disabled to enable the next rule
    '@typescript-eslint/no-unused-vars': ['error']
  }
}
```

```TS
export namespace foo.bar {
  export interface User {
    name: string;
  }
}
```

**Expected Result**

No lint errors.

**Actual Result**

```
error  'bar' is defined but never used  @typescript-eslint/no-unused-vars
```

**Additional Info**

Following doesn't trigger the error as far as I can tell.

```ts
export namespace foo {
  export namespace bar {
    export interface User {
      name: string;
    }
  }
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.0` |
| `@typescript-eslint/parser`        | `4.1.0` |
| `TypeScript`                       | `3.7.0` |
| `ESLint`                           | `7.9.0` |
| `node`                             | `10.22.0` |

**Comment by bradzacher** [comment]
I'm unable to repro this against master.
Likely you have old version of our tooling installed.

Could you please check the versions you have installed using one of the following?

```bash
$ npm list @typescript-eslint/eslint-plugin @typescript-eslint/parser
$ yarn list @typescript-eslint/eslint-plugin @typescript-eslint/parser
```

**Comment by hiranya911** [comment]
```
├── @typescript-eslint/eslint-plugin@4.1.1 
└── @typescript-eslint/parser@4.1.1 
```

I have several files with the above nested namespace pattern, and I get lint errors for all of them.

**Comment by bradzacher** [comment]
oops. I realised that I only tested the working example, sorry.
I can repro this against master.

**Comment by bradzacher** [comment]
So this is a rather big bug.
Our parser does not handle this case well, and our types are quite wrong.

`namespace foo.bar {}` produces the following AST:
`TSModuleDeclaration[id = foo] > TSModuleDeclaration[id = bar] > TSModuleBlock`
Really this should produce:
`TSModuleDeclaration[id = foo.bar] > TSModuleBlock`

I don't think anyone ever realised that TS would produce such a funky AST for this..
(I know I didn't know this was valid syntax until now!)

**Comment by hiranya911** [comment]
Interesting. Any idea how it handles the explicitly nested case: `namespace foo { namespace bar {}}`?

I'd assume it produces the same AST as in `namespace foo.bar {}`.

**Comment by bradzacher** [comment]
For `namespace foo { namespace bar {}}` we produce the following ast:
`TSModuleDeclaration[id = foo] > TSModuleBlock > TSModuleDeclaration[id = bar] > TSModuleBlock`

The AST we produce is driven by the AST TS produces. This is why it's so weird.
I guess they produce `TSModuleDeclaration[id = foo] > TSModuleDeclaration[id = bar]` because it simplifies some of their declaration merging logic?

For us we don't care about declaration merging so and care more about more closely representing the code structure, so we should really deviate from TS here.

**Comment by Guillaume-Mayer** [comment]
Hello,

Not exactly the same issue, but seems related.

**index.d.ts**
```ts
declare namespace A {
  export interface A {
    foo: string;
  }
}
```

Give me this warning: **'A' is defined but never used.**, on the namespace (even if it's used).

**Versions**
```json
    "@typescript-eslint/eslint-plugin": "4.2.0",
    "@typescript-eslint/parser": "4.2.0",
    "eslint": "7.9.0",
    "typescript": "4.0.3"
```

**Comment by bradzacher** [comment]
Entirely unrelated. Please open a new issue.

**Comment by Guillaume-Mayer** [comment]
#2595

**Comment by NoriSte** [comment]
This is the workaround I came up with: leveraging [varsIgnorePattern](https://eslint.org/docs/rules/no-unused-vars#varsignorepattern) in order to override the `@typescript-eslint/no-unused-vars` rule only for the `*.d.ts` files
```js
overrides: [
  {
    files: ['**/*.d.ts'],
    rules: {
      '@typescript-eslint/no-unused-vars': [
        'error',
        {
          varsIgnorePattern: 'bar|baz',
        },
      ],
    },
  },
],
```
Please note that this solution doesn't scale because you need to list all the parts of the domain name in the `varsIgnorePattern` pattern.

In our case we have a lot of `uui.domain.ui.list`, `uui.domain.actions.rpc.driver`, etc. namespaces and that's why we prefer to go with a 
```js
// @see https://github.com/typescript-eslint/typescript-eslint/issues/2573
// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
comment for every namespace we declare.

**Comment by armano2** [comment]
i looked into this and there are 3 ways to fix it:
1. change ast structure -> change that id is no longer an identifier but qualified name
2. disable creation of references in scope analysis for nested module declarations
3. modify no unused vars to mark variable in this syntax as used
   ```
       // module declaration in module declaration should not report unused vars error
      'TSModuleDeclaration > TSModuleDeclaration'(
        node: TSModuleDeclaration,
      ): void {
        if (node.id.type === AST_NODE_TYPES.Identifier) {
          let scope = context.getScope();
          if (scope.variableScope !== scope) {
            scope = scope.variableScope;
          } else if (scope.upper) {
            scope = scope.upper;
          }
          const superVar = scope.set.get(node.id.name);
          if (superVar) {
            superVar.eslintUsed = true;
          }
        }
      },
   ```

I'm kinda unsure what approach we want/should take this, @bradzacher what do you think?

**Comment by bradzacher** [comment]
short term hack would be (3).

proper solution would be (1).
This would also require a scope analyser change so it understands that `namespace X.Y {}` creates a type/value variable `X`.
